### PR TITLE
Add uniform_sample option to VideoReaderDecoder

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:
  New feature (non-breaking change which adds functionality)

  ## Description:
  Adds a `uniform_sample` boolean argument to `fn.experimental.readers.video` (issue #5926).
  When enabled, exactly `sequence_length` frames are sampled uniformly across the full video
  (or the ROI defined by `file_list`), using indices equivalent to
  `numpy.linspace(start, end-1, sequence_length)` rounded to the nearest integer
  (ties round away from zero, matching C++ std::round).

  Each video produces exactly one sample per epoch regardless of video length.
  The `stride` and `step` arguments are ignored when `uniform_sample=True`.

  ## Additional information:

  ### Affected modules and functionalities:
  - dali/operators/video/reader/video_reader_decoder_op.cc: added `uniform_sample` schema
    argument; extended VideoSampleDesc with an explicit frame index list; added uniform
    sampling branch in PrepareMetadataImpl; updated Prefetch to dispatch to
    DecodeFrames(span<const int>) when explicit indices are present.
  - fn.experimental.readers.video Python API: new `uniform_sample` optional argument
    (default False), documented in the schema docstring.

  ### Key points relevant for the review:
  - When uniform_sample=True, VideoSampleDesc::frame_idxs_ is non-empty and drives decoding
    instead of start_/end_/stride_. All existing stride-based paths are unchanged.
  - The sequence_length=1 edge case yields frame 0 (linspace of a single element is [start]).
  - ROI from file_list is respected: linspace is computed over [start_frame, end_frame).
  - enable_frame_num="sequence" correctly reports the sampled absolute frame indices.
  - enable_timestamps works as before (timestamps populated per decoded frame).

  ### Tests:
  - [x] New tests added
    - [x] Python tests: dali/test/python/test_video_reader.py::test_uniform_sample
      - Parametrized over device in {cpu, gpu} x sequence_length in {5, 1}
      - Verifies epoch_size == num_video_files (one sample per video)
      - Verifies frame indices match linspace with C++ rounding semantics, using the same
        DALI decoder (stride=1 reader) as ground truth — no reliance on cv2 metadata
      - GPU: pixel-level comparison against a full-video decode; CPU skipped (libavcodec
        seek inaccuracy with H.264 B-frames is a pre-existing decoder limitation,
        not a regression)

  ## Checklist

  ### Documentation
  - [x] Documentation updated
    - [x] Docstring (schema AddOptionalArg docstring in video_reader_decoder_op.cc)

  ### DALI team only

  #### Requirements
  - [x] Implements new requirements

  REQ IDs: N/A
  JIRA TASK: DALI-4273

### CODE REVIEW

**Comment by jantonguirao** [comment]
!build

File: dali/operators/video/reader/video_reader_decoder_op.cc [hunk 1]
```diff
@@ -76,11 +76,12 @@ class FrameIndexCache {
 
 struct VideoSampleDesc {
   VideoSampleDesc(const VideoFileMeta *video_file_meta = nullptr, int start = -1, int end = -1, int stride = -1)
-      : video_file_meta_(video_file_meta), start_(start), end_(end), stride_(stride) {}
+      : video_file_meta_(video_file_meta), start_(start), end_(end), stride_(stride), frame_idxs_{} {}
```
Comment (mdabek-nvidia) [inline dali/operators/video/reader/video_reader_decoder_op.cc]:
```suggestion
      : video_file_meta_(video_file_meta), start_(start), end_(end), stride_(stride), frame_idxs_() {}
```

Nitpick, to keep the style consistent

File: dali/test/python/test_video_reader.py [hunk 2]
```diff
@@ -241,3 +242,142 @@ def test_compare_experimental_to_legacy_reader_file_root(
         pad_mode=pad_mode,
         image_type=image_type,
     )
+
+
+@cartesian_params(
+    devices,
+    [5, 1],  # sequence lengths including edge case N=1
```
Comment (mdabek-nvidia) [inline dali/test/python/test_video_reader.py:249]: Do we check what happens if the sequence_length == 0? I am not sure if this should be an error or just a pass through.

File: dali/test/python/test_video_reader.py [hunk 3]
```diff
@@ -241,3 +242,142 @@ def test_compare_experimental_to_legacy_reader_file_root(
...
+
+    # Get per-video frame counts and (GPU only) decode all frames for pixel comparison.
+    # Iterating seq_len=1 gives both in one pass; CPU only needs the count via get_metadata().
+    # GPU only: NVDEC produces identical pixels whether seeking or decoding sequentially.
+    # CPU (libavcodec) can produce minor seek-induced differences for H.264 B-frames.
+    frame_counts = []
+    all_frames = []  # populated on GPU only
+    for video_file in video_files_sorted:
+        reader = ndd.experimental.readers.Video(
+            device=device, filenames=[video_file], sequence_length=1, stride=1, step=1
```
Comment (mdabek-nvidia) [inline dali/test/python/test_video_reader.py:263]: According to the following warnings: [dali/operators/video/reader/video_reader_decoder_op.cc](https://github.com/NVIDIA/DALI/pull/6258/changes#diff-2feeaead14638efe39357646a4bebbeb2c4393678347cf2ef493229e182ce5a1), stride and step should be ignored. Do we check it ?

File: dali/test/python/test_video_reader.py [hunk 4]
```diff
@@ -241,3 +242,198 @@ def test_compare_experimental_to_legacy_reader_file_root(
...
+    video_file = sorted(VIDEO_FILES)[0]
+    sequence_length = 5
+
+    reader_default = ndd.experimental.readers.Video(
+        device=device,
+        filenames=[video_file],
+        sequence_length=sequence_length,
+        uniform_sample=True,
+        enable_frame_num="sequence",
+    )
```
Comment (stiepan) [inline dali/test/python/test_video_reader.py:415]: just wanted to say that ndd looks like a game-changer for writing DALI tests :D

File: dali/operators/video/reader/video_reader_decoder_op.cc [hunk 5]
```diff
@@ -306,20 +326,37 @@ class VideoLoaderDecoder : public Loader<Backend, Sample, true> {
...
-        samples_.emplace_back(&entry, start, start + full_seq_stride, stride_);
-      }
+      if (uniform_sample_) {
+        int total_frames = entry.end_frame - entry.start_frame;
+        VideoSampleDesc s(&entry, entry.start_frame, entry.end_frame, 1);
+        all_frame_idxs_.emplace_back(sequence_len_);
+        auto& idxs = all_frame_idxs_.back();
+        for (int i = 0; i < sequence_len_; ++i) {
+          double t = (sequence_len_ > 1) ? (double)i / (sequence_len_ - 1) : 0.0;
+          idxs[i] = entry.start_frame + static_cast<int>(std::round(t * (total_frames - 1)));
```
Comment (stiepan) [inline dali/operators/video/reader/video_reader_decoder_op.cc:336]: Just to double check, the `entry.end_frame` is not a valid index of a last frame in the video but sth like past one?

```
idxs[sequence_len - 1] = start_frame + int(round(1 * (total_frames - 1))) = start_frame + total_frames - 1 = start_frame + end_frame - start_frame - 1 = end_frame  - 1
```
